subsys/net/lib/aws_jobs/src/aws_jobs.c

```diff
@@ -0,0 +1,397 @@
+/*
+ *copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <zephyr.h>
+#include <stdio.h>
+#include <net/mqtt.h>
+#include <logging/log.h>
+#include "aws_jobs.h"
+
+LOG_MODULE_REGISTER(aws_jobs, CONFIG_AWS_JOBS_LOG_LEVEL);
+
+
+/**
+ * @brief Local function used to error check snprintf outputs.
+ *
+ * @param[in] ret Return value of snpritnf.
+ * @param[in] size Max size of what would have been the final string.
+ * @param[in] entry Name of the function it was called from.
+ *
+ * @retval 0 If sucessful
+ * @retval -ENOMEM If the buffer was too small for the string. Otherwise
+ *	the error code of snprintf if the formating failed.
+ */
+static inline int report_snprintf_err(u32_t ret, u32_t size, const char *entry)
+{
+	if (ret >= size) {
+		LOG_ERR("Unable to fit formated string into to allocate "
+			"memory for %s", entry);
+		return -ENOMEM;
+	} else if (ret < 0) {
+		LOG_ERR("Output error for %s was encountered with return value "
+			"%d", entry, ret);
+		return ret;
+	}
+	return 0;
+}
+
+
+/**
+ * @brief Constructs the get topic for accepted and rejeceted AWS jobs.
+ *
+ * @param[in] client_id Client ID of the current connected MQTT session.
+ * @param[out] topic Outputs the constructed mqtt_topic into the given buffer.
+ *
+ * @retval 0 If sucessful otherwise the return value of report_snprintf_err
+ */
+static int construct_get_topic(const u8_t *client_id, struct mqtt_topic *topic)
+{
+	u32_t get_topic_len;
+	u8_t get_topic[GET_TOPIC_LEN + 1];
+	int ret = snprintf(get_topic,
+			   GET_TOPIC_LEN,
+			   GET_TOPIC_TEMPLATE,
+			   client_id,
+			   "#");
+	get_topic_len = ret + 1;
+	ret = report_snprintf_err(ret, GET_TOPIC_LEN, "get_topic");
+
+	if (ret) {
+		return ret;
+	}
+
+	topic->topic.utf8 = get_topic;
+	topic->topic.size = get_topic_len;
+	topic->qos = MQTT_QOS_1_AT_LEAST_ONCE;
+	return 0;
+}
+
+
+int aws_jobs_subscribe_get(struct mqtt_client *const client)
+{
+	__ASSERT(client != NULL, "invalid client");
```

Comment:
Top-level API error checking should return an error instead of asserting.

Reply:
I'll add an `-EINVAL` return for all top-level API calls instead of asserts